### PR TITLE
Updating optimizer to handle arrays with nint indexes

### LINKED ISSUE #44415: Bounds checks are no longer elided when using nint for indexing.
Examine the following two methods that only differ based on using `int` vs `nint` for indexing into an array:
```csharp
public static int SumWithInt32Index(int[] a)
{
    int result = 0;

    for (int i = 0; i < a.Length; i++)
    {
        result += a[i];
    }

    return result;
}

public static int SumWithNativeIntIndex(int[] a)
{
    int result = 0;
        
    for (nint i = 0; i < a.Length; i++)
    {
        result += a[i];
    }
        
    return result;
}
```

These two methods result in the following codegen:
```asm
; ConsoleApp64.Benchmarks.SumWithInt32Index(Int32[])
       xor       eax,eax
       xor       edx,edx
       mov       r8d,[rcx+8]
       test      r8d,r8d
       jle       short M01_L01
M01_L00:
       movsxd    r9,edx
       add       eax,[rcx+r9*4+10]
       inc       edx
       cmp       r8d,edx
       jg        short M01_L00
M01_L01:
       ret
; Total bytes of code 29

; ConsoleApp64.Benchmarks.SumWithNativeIntIndex(Int32[])
       sub       rsp,28
       xor       eax,eax
       xor       edx,edx
       mov       r8d,[rcx+8]
       movsxd    r8,r8d
       test      r8,r8
       jle       short M01_L01
M01_L00:
       cmp       rdx,r8
       jae       short M01_L02
       add       eax,[rcx+rdx*4+10]
       inc       rdx
       cmp       r8,rdx
       jg        short M01_L00
M01_L01:
       add       rsp,28
       ret
M01_L02:
       call      CORINFO_HELP_RNGCHKFAIL
       int       3
; Total bytes of code 48
```

As can be seen, the latter is no longer able to elide the bounds checks due to the additional cast around `a.Length` to `nint` which results in significantly worse codegen. A developer must use `Unsafe.Add` to avoid the bounds checks in the inner loop and `MemoryMarshal.GetArrayDataReference` to avoid bounds checks in the outer loop. In both cases, while the inner loop does improve, the overall codegen is still slightly worse:
```asm
; ConsoleApp64.Benchmarks.SumWithNativeIntIndex(Int32[])
       xor       eax,eax
       lea       rdx,[rcx+10]
       xor       r8d,r8d
       mov       ecx,[rcx+8]
       movsxd    rcx,ecx
       test      rcx,rcx
       jle       short M01_L01
M01_L00:
       add       eax,[rdx+r8*4]
       inc       r8
       cmp       rcx,r8
       jg        short M01_L00
M01_L01:
       ret
; Total bytes of code 33
```

It would be beneficial if utilizing `nint` where it is accepted "just" works.

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/runtime/issues/44415

[JitDump.txt](https://github.com/dotnet/runtime/files/5877200/JitDump.txt) for a simple console app:
```csharp
using System;
using System.Runtime.CompilerServices;

class C
{
    static void Main()
    {
        int[] a = new int[] { 1 };
        int r = F(a);
        Console.WriteLine(r);
    }

    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    static int F(int[] a)
    {
        int result = 0;

        for (nint i = 0; i < a.Length; i++)
        {
            result += a[i];
        }

        return result;
    }
}
```

Shows:
```asm
IN0001: 000000 xor      eax, eax
IN0002: 000002 xor      rdx, rdx
IN0003: 000004 mov      r8d, dword ptr [rcx+8]
IN0004: 000008 movsxd   r8, r8d
IN0005: 00000B test     r8, r8
IN0006: 00000E jle      SHORT G_M64818_IG04
IN0007: 000010 align    

G_M64818_IG03:        ; offs=000010H, size=000CH, bbWeight=4    PerfScore 14.00, gcrefRegs=00000002 {rcx}, byrefRegs=00000000 {}, byref, isz

IN0008: 000010 add      eax, dword ptr [rcx+4*rdx+16]
IN0009: 000014 inc      rdx
IN000a: 000017 cmp      r8, rdx
IN000b: 00001A jg       SHORT G_M64818_IG03

G_M64818_IG04:        ; offs=00001CH, size=0001H, bbWeight=1    PerfScore 1.00, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, epilog, nogc

IN000c: 00001C ret      
```

### CODE REVIEW

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -718,7 +718,7 @@ bool Compiler::optPopulateInitInfo(unsigned loopInd, GenTree* init, unsigned ite
 
     // RHS can be constant or local var.
     // TODO-CQ: CLONE: Add arr length for descending loops.
-    if (rhs->gtOper == GT_CNS_INT && rhs->TypeGet() == TYP_INT)
+    if (rhs->gtOper == GT_CNS_INT && ((rhs->TypeGet() == TYP_INT) || (rhs->TypeGet() == TYP_I_IMPL)))
```
Comment (tannergooding) [inline src/coreclr/jit/optimizer.cpp:721]: This works by allowing either `TYP_INT` or `TYP_I_IMPL`
In the case of `TYP_I_IMPL` it also looks for `GT_CAST` to that type and gets the underlying node when that exists

**Comment by AndyAyersMS** [comment]
cc @BruceForstall 

Seems plausible?  I am not that familiar with all the consumers of `optExtractArrIndex` so it's probably worth looking them over to make sure they're not going to be surprised.

**Comment by AndyAyersMS** [comment]
Also FWIW I prefer seeing `OperIs` and `TypeIs` as I find they're more readable...

**Comment by omariom** [comment]
@tannergooding  Will it work for Spans?

**Comment by tannergooding** [comment]
> Will it work for Spans?

@omariom, no because `Span<T>` doesn't have an indexer that takes anything other than `int`.

**Comment by omariom** [comment]
@tannergooding  Then may be it is good opportunity to introduce one? :)

**Comment by tannergooding** [comment]
> Then may be it is good opportunity to introduce one? :)

@omariom, feel free to open an API proposal for it: https://github.com/dotnet/runtime/issues/new?assignees=&labels=api-suggestion&template=02_api_proposal.md.
The relevant area owners can then comment and it can go to API review if they agree it is a good idea.

**Comment by tannergooding** [comment]
Figured out the failure (was reassigning `index` when I should have reassigned `indexVar`) and the current commit has zero diffs for `--pmi --frameworks`:
```
Found 271 files with textual diffs.

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 51164958
Total bytes of diff: 51164958
Total bytes of delta: 0 (0.00% of base)


0 total files with Code Size differences (0 improved, 0 regressed), 270 unchanged.

0 total methods with Code Size differences (0 improved, 0 regressed), 263288 unchanged.

--------------------------------------------------------------------------------
```

It continues to handle the case in the original post around using `nuint`.

**Comment by tannergooding** [comment]
CC. @dotnet/jit-contrib, @BruceForstall this should be ready for review

**Comment by briansull** [comment]
Yes, I believe that  this approach of updating the loop optimizer to recognize the pattern will work

**Comment by tannergooding** [comment]
Fixed so there are zero-diffs for x86 as well. Turns out there are some `GT_CAST: TYP_INT` nodes that get picked up on x86. It might be interesting to look at these for a different PR, but I felt it was better to not include in this one.

Likewise, there are some casts that it looks like it would be good to handle that aren't today. Namely, when handling `nint/nuint` for array indexing C# generates the following today (where `loc.0` is `nint`):
```
ldloc.0
conv.i8
conv.ovf.i
```

This is somewhat "problematic" because on 64-bit we get a `long->long.ovfl` cast node that is never trimmed because it is marked as "overflow", even though it can't possibly overflow (as the type is the same).

Likewise for x86, this means we get a `int->long->int.ovfl` casting nodes, which likewise can't overflow because the sign isn't changing and so the value is preserved across the upcast and subsequent downcast.

I also left these to some "future" PR.

**Comment by tannergooding** [comment]
Going to close this as it looks like there is some additional work to be done here to determine what's causing the failures.

**Pull request closed without merging**